File: src/APL/types/dimensions/DervDimDop.java
package APL.types.dimensions;

import APL.Scope;
import APL.errors.SyntaxError;
import APL.types.*;
import APL.types.functions.*;

public class DervDimDop extends Dop {
  private final Dop f;
  private final int dim;
  
  @Override public String repr() {
    return f.repr()+"["+dim+"]";
  }
  
  public DervDimDop(Dop f, Integer dim, Scope sc) {
    super(sc);
    this.f = f;
    if (dim == null) this.dim = 0;
    else if (dim < 0) this.dim = dim;
    else this.dim = dim - sc.IO;
    this.token = f.token;
    
  }
  
  @Override
  public Value call(Obj aa, Obj ww, Value a, Value w, DerivedDop derv) {
    if (!(f instanceof DimDDop)) throw new SyntaxError("Attempt to call function dyadically that doesn't support dimension specification", a);
    return ((DimDDop) f).call(aa, ww, a, w, dim);
  }
  
  @Override
  public Value call(Obj aa, Obj ww, Value w, DerivedDop derv) {
    if (!(f instanceof DimMDop)) throw new SyntaxError("Attempt to call function monadically that doesn't support dimension specification", w);
    return ((DimMDop) f).call(aa, ww, w, dim);
  }
}

File: APL/types/dimensions/DervDimDop.cpp
#include <APL/DervDimDop.h>
#include <APL/DimDDop.h>
#include <APL/DimMDop.h>

namespace APL::types::dimensions
{
    using Scope = APL::Scope;
    using SyntaxError = APL::errors::SyntaxError;
    // using namespace APL::types;
    // using namespace APL::types::functions;

    std::string DervDimDop::repr()
    {
      return f->repr() + L"[" + std::to_wstring(dim) + L"]";
    }

    DervDimDop::DervDimDop(std::shared_ptr<Dop> f, std::optional<int> &dim, std::shared_ptr<Scope> sc) : Dop(sc), f(f)
    {
      if (!dim)
      {
          this->dim = 0;
      }
      else if (dim < 0)
      {
          this->dim = dim.value();
      }
      else
      {
          this->dim = dim.value() - sc->IO;
      }
      this->token = f->token;

    }

    std::shared_ptr<APL::types::Value> DervDimDop::call(std::shared_ptr<APL::types::Obj> aa, std::shared_ptr<APL::types::Obj> ww, std::shared_ptr<APL::types::Value> a, std::shared_ptr<APL::types::Value> w, std::shared_ptr<DerivedDop> derv)
    {
      if (!(std::dynamic_pointer_cast<DimDDop>(f) != nullptr))
      {
          throw SyntaxError(L"Attempt to call function dyadically that doesn't support dimension specification", a);
      }
      return (std::static_pointer_cast<DimDDop>(f))->call(aa, ww, a, w, dim);
    }

    std::shared_ptr<APL::types::Value> DervDimDop::call(std::shared_ptr<APL::types::Obj> aa, std::shared_ptr<APL::types::Obj> ww, std::shared_ptr<APL::types::Value> w, std::shared_ptr<DerivedDop> derv)
    {
      if (!(std::dynamic_pointer_cast<DimMDop>(f) != nullptr))
      {
          throw SyntaxError(L"Attempt to call function monadically that doesn't support dimension specification", w);
      }
      return (std::static_pointer_cast<DimMDop>(f))->call(aa, ww, w, dim);
    }
}
